src/backends/torch/torchdataset.cc

```diff
@@ -311,4 +372,131 @@ namespace dd
     new_dataset._batches.insert(new_dataset._batches.end(), start_it, stop_it);
     return new_dataset;
   }
+
+  /*-- image tools --*/
+  int TorchDataset::add_image_file(const std::string &fname, const int &target,
+                                   const int &height, const int &width)
+  {
+    ImgTorchInputFileConn *inputc
+        = reinterpret_cast<ImgTorchInputFileConn *>(_inputc);
+
+    DDImg dimg;
+    inputc->copy_parameters_to(dimg);
+
+    try
+      {
+        if (dimg.read_file(fname))
+          {
+            this->_logger->error("Uri failed: {}", fname);
+          }
+      }
+    catch (std::exception &e)
+      {
+        this->_logger->error("Uri failed: {}", fname);
+      }
+    if (dimg._imgs.size() != 0)
+      {
+        at::Tensor imgt = image_to_tensor(dimg._imgs[0], height, width);
+        at::Tensor targett = target_to_tensor(target);
+
+        add_batch({ imgt }, { targett });
+        return 0;
+      }
+    else
+      {
+        return -1;
+      }
+  }
+
+  int TorchDataset::add_image_file(const std::string &fname,
+                                   const std::vector<double> &target,
+                                   const int &height, const int &width)
+  {
+    ImgTorchInputFileConn *inputc
+        = reinterpret_cast<ImgTorchInputFileConn *>(_inputc);
+
+    DDImg dimg;
+    inputc->copy_parameters_to(dimg);
+
+    try
+      {
+        if (dimg.read_file(fname))
+          {
+            this->_logger->error("Uri failed: {}", fname);
+          }
+      }
+    catch (std::exception &e)
+      {
+        this->_logger->error("Uri failed: {}", fname);
+      }
+    if (dimg._imgs.size() != 0)
+      {
+        at::Tensor imgt = image_to_tensor(dimg._imgs[0], height, width);
+        // at::Tensor targett{ torch::full(1, target, torch::kLong) };
+        at::Tensor targett = target_to_tensor(target);
+
+        add_batch({ imgt }, { targett });
+        return 0;
+      }
+    else
+      {
+        return -1;
+      }
+  }
+
+  at::Tensor TorchDataset::image_to_tensor(const cv::Mat &bgr,
```

Comment:
`image_to_tensor` is a classical method because it uses attributes of `_inputc` (mean, scale std...),
=>  so you need to add `_inputc` as an attribute of dataset so that `image_to_tensor` can use all these. 

I wonder of it would not be simpler to make `image_to_tensor` function a generic util function ie `torch_utils::image_to_tensor`, and passing all required parameters explicitly to it would allow not to  put `inputc` as an attribute of dataset and would make more explicit what is needed to transform an image to a tensor

(or if you do not want to add all parameters, just pass a reference to inputc; ideally it should be something in between : a struct in the inputc that contains the transformation parameters needed )

Reply:
Unfortunately `inputc` is needed in forthcoming PRs when encoded images are stored in dbs (instead of tensors). This is because `DDimg` and `inputc->copy_parameters_to` are then needed. Agreed that this cannot be seen from this PR alone.

There are other forthcoming issues of this kind coming up.

Unless you and others think otherwise, I do favor to get it working then rationalize / simplify as best as we can once all features are in. 

Basically have dataset on the side of the input connector does run us into unexpected issues.